### PR TITLE
Send remainder polynomial instead of remainder code word

### PR DESCRIPTION
**Pull request by Al-Kindi-0**
This PR proposes another method for performing the final check in the FRI query stage. Specifically, in the current implementation the verifier receives a code word from the prover which the latter claims to be consistent with an honest execution of the protocol.
The verifier then checks that the commitment to the received code word matches the one received in the commit phase of the protocol. The verifier then goes on to interpolate the code word in order to find the corresponding remainder polynomial. The verifier concludes by checking whether the degree of this polynomial is less than the expected degree bound (which can be computed from the input polynomial degree bound and the protocol parameters).
The current proposal simplifies things by sending the remainder polynomial directly. Thus the verifier can perform the degree check almost trivially. Moreover, no commitments to the remainder polynomial are needed.

### CODE REVIEW

File: fri/src/prover/mod.rs [hunk 1]
```diff
@@ -222,12 +213,13 @@ where
         });
     }
 
-    /// Creates a FriRemainder from a vector of `evaluations` representing the remainder and
-    /// commits to the latter by hashing sequentially `evaluations`.
-    fn set_remainder(&mut self, channel: &mut C, evaluations: &mut [E]) {
-        let commitment = <H as ElementHasher>::hash_elements(evaluations);
-        channel.commit_fri_layer(commitment);
-        self.remainder = FriRemainder(evaluations.to_vec());
+    /// Creates remainder polynomial from a vector of `evaluations` representing the remainder
```
Comment (Nashtare) [inline fri/src/prover/mod.rs:216]: I'd perhaps reword this to emphasize that we create a poly in coeff form, from its evaluations. But not a big deal

File: fri/src/verifier/mod.rs [hunk 2]
```diff
@@ -378,3 +346,11 @@ fn get_query_values<E: FieldElement, const N: usize>(
 
     result
 }
+
+// Evaluates a polynomial with coefficients in an extension field at a point in the base field.
+pub fn eval_horner<E>(p: &[E], x: E) -> E
```
Comment (Nashtare) [inline fri/src/verifier/mod.rs:351]: The description of the method doesn't match it's actual behavior, as both `p` coefficients and the target point `x` are in the same field `E`. Plus when we're calling the horner evaluation in the verification, we're converting `x` from `E::BaseField` to `E`.

**Review by Nashtare — COMMENTED** [review]
I took the liberty to add some light comments.
Also, I think you forgot to update in `examples/src/lib.rs` the default parameters of `to_proof_options()` (it's still using 256 for the now `fri_remainder_size` field.

File: air/src/options.rs [hunk 3]
```diff
@@ -89,15 +89,15 @@ impl ProofOptions {
     /// * `blowup_factor` is smaller than 4, greater than 256, or is not a power of two.
     /// * `grinding_factor` is greater than 32.
     /// * `fri_folding_factor` is not 4, 8, or 16.
-    /// * `fri_max_remainder_size` is smaller than 32, greater than 1024, or is not a power of two.
+    /// * `fri_remainder_size` is smaller than 1, greater than 32, or is not a power of two.
```
[the remark this answers is not in the document]
Comment (Al-Kindi-0) [reply air/src/options.rs:92]: Done!

File: fri/src/verifier/mod.rs (hunk 2 above)
Comment (Al-Kindi-0) [reply fri/src/verifier/mod.rs:351]: Thank you! I updated it to be more consistent with my original way of thinking. Let me know if it works now.

File: fri/src/verifier/mod.rs [hunk 4]
```diff
@@ -378,3 +346,14 @@ fn get_query_values<E: FieldElement, const N: usize>(
 
     result
 }
+
+// Evaluates a polynomial with coefficients in an extension field at a point in the base field.
+pub fn eval_horner<E, B>(p: &[E], x: B) -> E
+where
+    E: FieldElement<BaseField = B>,
+    B: FieldElement,
```
Comment (Nashtare) [inline fri/src/verifier/mod.rs:354]: Couldn't we abstract the second type and refer to `E::BaseField`? Something like:

```suggestion
pub fn eval_horner<E>(p: &[E], x: E::BaseField) -> E
where
    E: FieldElement,
```

File: fri/src/verifier/mod.rs (hunk 2 above)
Comment (Nashtare) [reply fri/src/verifier/mod.rs]: Thanks! I've added a comment, I think we can go down to a single generic parameter to make the function definition lighter.

Comment (Al-Kindi-0) [reply fri/src/verifier/mod.rs]: Thanks @Nashtare ! I looks cleaner indeed.

Comment (Nashtare) [reply fri/src/verifier/mod.rs]: Sorry for breaking the build, we need to update also when the function is called, to remove the extra `B` 😅

Comment (Al-Kindi-0) [reply fri/src/verifier/mod.rs]: Sure thing, I updated it now :-)

**Comment by Al-Kindi-0** [comment]
Thank you for the review!

File: fri/src/verifier/mod.rs [hunk 5]
```diff
@@ -303,56 +303,24 @@ where
...
         for (&position, evaluation) in positions.iter().zip(evaluations) {
-            if remainder[position] != evaluation {
+            let comp_eval = eval_horner::<E>(
+                &remainder,
+                offset * domain_generator.exp((position as u64).into()),
+            );
+            if comp_eval != evaluation {
                 return Err(VerifierError::InvalidRemainderFolding);
             }
         }
```
Comment (irakliyk) [inline fri/src/verifier/mod.rs:321]: A couple of small nits here:
- I'd probably rename `remainder` into `remainder_poly` to indicate that it is a polynomial in coefficient form.
- I'd use `exp_vartime()` instead of `exp()` here to make exponentiation potentially faster.

File: fri/src/verifier/channel.rs [hunk 6]
```diff
@@ -90,26 +90,17 @@ pub trait VerifierChannel<E: FieldElement> {
...
-
-        // sequential hash of the remainder
-        let set_commitment =
-            <<Self as VerifierChannel<E>>::Hasher as ElementHasher>::hash_elements(&remainder);
-
-        if *commitment != set_commitment {
-            return Err(VerifierError::RemainderCommitmentMismatch);
+        if polynom::degree_of(&remainder) > max_degree {
+            return Err(VerifierError::RemainderDegreeMismatch(max_degree));
         }
```
Comment (irakliyk) [inline fri/src/verifier/channel.rs:104]: I think it would be cleaner to perform this check in the verifier (i.e., in `verify_generic()` function).

Also, can we simplify this as:
```Rust
if remainder.len() > max_degree + 1 {
    return Err(VerifierError::RemainderDegreeMismatch(max_degree));
}
```

File: air/src/options.rs [hunk 7]
```diff
@@ -63,7 +63,7 @@ pub struct ProofOptions {
     grinding_factor: u8,
     field_extension: FieldExtension,
     fri_folding_factor: u8,
-    fri_max_remainder_size: u8, // stored as power of 2
```
Comment (irakliyk) [inline air/src/options.rs:66]: Should we name this `fri_max_remainder_degree`? I think this would better reflect what it actually is. We could still store it as a power of two, but now we'd be storing `log2(max_degree + 1)`.

File: air/src/options.rs [hunk 8]
```diff
@@ -114,17 +115,17 @@ impl ProofOptions {
...
+        assert!(fri_remainder_size >= 1, "FRI polynomial remainder size cannot be 0");
+        assert!(fri_remainder_size <= 32, "FRI polynomial remainder size cannot be greater than 32");
 
         ProofOptions {
             num_queries: num_queries as u8,
             blowup_factor: blowup_factor as u8,
             grinding_factor: grinding_factor as u8,
             field_extension,
             fri_folding_factor: fri_folding_factor as u8,
-            fri_max_remainder_size: fri_max_remainder_size.trailing_zeros() as u8,
```
Comment (irakliyk) [inline air/src/options.rs:127]: If we store remainder degree, this would be `log2(fri_max_remainder_degree + 1)`.

File: fri/src/options.rs [hunk 9]
```diff
@@ -12,7 +12,7 @@ use math::StarkField;
 #[derive(Clone, PartialEq, Eq)]
 pub struct FriOptions {
     folding_factor: usize,
-    max_remainder_size: usize,
+    remainder_size: usize,
```
Comment (irakliyk) [inline fri/src/options.rs:15]: Similar comment to the one before: should we rename this into `max_remainder_degree`?

File: fri/src/prover/mod.rs [hunk 10]
```diff
@@ -222,12 +213,12 @@ where
...
-        let commitment = <H as ElementHasher>::hash_elements(evaluations);
-        channel.commit_fri_layer(commitment);
-        self.remainder = FriRemainder(evaluations.to_vec());
+    /// Creates remainder polynomial in coefficient form from a vector of `evaluations` over a domain.
+    fn set_remainder(&mut self, evaluations: &mut [E]) {
+        let inv_twiddles = fft::get_inv_twiddles(evaluations.len());
+        fft::interpolate_poly_with_offset(evaluations, &inv_twiddles, E::BaseField::GENERATOR);
+        let remainder = evaluations[..evaluations.len() / self.options.blowup_factor()].to_vec();
+        self.remainder = FriRemainder(remainder);
     }
```
Comment (irakliyk) [inline fri/src/prover/mod.rs:222]: A few comments/questions here:
- Should we get the domain offset from options here rather than using `E::BaseField::GENERATOR`?
- One of the things that we did in `commit_fri_layer()` was reseeding of the random coin. I think we might still need to do this, but now we should reseed it with the entire remainder polynomial - right? If so, it might be more efficient to commit to the remainder polynomial and then use this commitment to reseed the public coin with this commitment.
- minor nit: I'd probably rename `remainder` variable into `remainder_poly`.

**Review by irakliyk — CHANGES_REQUESTED** [review]
Looks great! Thank you! I left several comments inline.

File: fri/src/verifier/mod.rs (hunk 5 above)
Comment (Al-Kindi-0) [reply fri/src/verifier/mod.rs:321]: Done!

File: fri/src/options.rs (hunk 9 above)
Comment (Al-Kindi-0) [reply fri/src/options.rs:15]: I have renamed all related variables as `max_remainder_degree_plus_1`. It is a bit verbose but accurate.

File: air/src/options.rs (hunk 7 above)
Comment (Al-Kindi-0) [reply air/src/options.rs:66]: Makes sense.

File: fri/src/verifier/channel.rs (hunk 6 above)
Comment (Al-Kindi-0) [reply fri/src/verifier/channel.rs:104]: Done!

File: fri/src/verifier/channel.rs [hunk 11]
```diff
@@ -90,28 +90,16 @@ pub trait VerifierChannel<E: FieldElement> {
         Ok(group_vector_elements(layer_queries))
     }
 
-    /// Returns FRI remainder values (last FRI layer) read from this channel.
+    /// Returns FRI remainder polynomial read from this channel.
     ///
-    /// This also checks whether the remainder is valid against the provided commitment.
+    /// This also
     ///
```
Comment (irakliyk) [inline fri/src/verifier/channel.rs:96]: nit: we should probably delete these lines.

File: fri/src/verifier/channel.rs [hunk 12]
```diff
@@ -90,28 +90,16 @@ pub trait VerifierChannel<E: FieldElement> {
...
     /// # Errors
     /// Returns an error if:
-    /// - If the sequrntial hash constructed from the remainder values does not match the specified
-    ///   `commitment`.
-    fn read_remainder<const N: usize>(
-        &mut self,
-        commitment: &<<Self as VerifierChannel<E>>::Hasher as Hasher>::Digest,
-    ) -> Result<Vec<E>, VerifierError> {
+    /// - If the degree of the polynomial is greater than the maximum expected degree.
```
Comment (irakliyk) [inline fri/src/verifier/channel.rs:99]: We should probably remove the "Errors" section.

File: fri/src/options.rs [hunk 13]
```diff
@@ -87,28 +84,14 @@ impl FriOptions {
...
     ///
-    /// The remainder layer (the last FRI layer) is not included in the returned value.
-    ///
     /// The number of layers for a given domain size is defined by the `folding_factor` and
-    /// `max_remainder_size` settings.
+    /// `remainder_max_degree_plus_1` settings.
     pub fn num_fri_layers(&self, mut domain_size: usize) -> usize {
         let mut result = 0;
-        while domain_size > self.max_remainder_size {
+        while domain_size > (self.remainder_max_degree_plus_1 * self.blowup_factor) {
```
Comment (irakliyk) [inline fri/src/options.rs:91]: minor nit: for clarity, I'd probably write this as:
```Rust
let max_remainder_size = self.remainder_max_degree_plus_1 * self.blowup_factor;
while domain_size < max_remainder_size {
```

File: fri/src/prover/mod.rs [hunk 14]
```diff
@@ -222,12 +214,15 @@ where
...
 
-    /// Creates a FriRemainder from a vector of `evaluations` representing the remainder and
-    /// commits to the latter by hashing sequentially `evaluations`.
+    /// Creates remainder polynomial in coefficient form from a vector of `evaluations` over a domain.
     fn set_remainder(&mut self, channel: &mut C, evaluations: &mut [E]) {
-        let commitment = <H as ElementHasher>::hash_elements(evaluations);
+        let inv_twiddles = fft::get_inv_twiddles(evaluations.len());
+        fft::interpolate_poly_with_offset(evaluations, &inv_twiddles, self.options.domain_offset());
+        let remainder_poly =
+            evaluations[..evaluations.len() / self.options.blowup_factor()].to_vec();
```
Comment (irakliyk) [inline fri/src/prover/mod.rs:222]: minor nit: for clarity, I'd do it like this:
```Rust
let remainder_size = evaluations.len() / self.options.blowup_factor();
let remainder_poly = evaluations[..remainder_size].to_vec();
```

**Review by irakliyk — CHANGES_REQUESTED** [review]
Looks great! Thank you! Just a few more minor nits.

**Review by irakliyk — APPROVED** [review]
All looks good! Thank you!
